torch/csrc/jit/type.cpp

```diff
@@ -248,10 +255,20 @@ TypePtr matchTypeVariables(TypePtr formal, TypePtr actual, TypeEnv& type_env) {
   } else if (auto opt_formal = formal->cast<OptionalType>()) {
     if (auto opt_actual = actual->cast<OptionalType>()) {
       return OptionalType::create(matchTypeVariables(
-          opt_formal->getElementType(), opt_actual->getElementType(), type_env));
+          opt_formal->getElementType(),
+          opt_actual->getElementType(),
+          type_env));
     } else {
+      if (opt_formal->getElementType()->kind() == TypeKind::VarType) {
```

Comment:
Do you mind commenting what this specifies?

(Optional[VarType], None) -> None, 
(Optional[Vartype], T) -> T , 
(Optional[Vartype], Optional[T1]) -> Optional[T1]

And add a test in test_jit.py for each case ? and add a test for unifying T, None across different branches in an if statement?